src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encoder;
+
+/**
+ * Hashes passwords using the best available encoder.
+ * Validates them using a chain of encoders.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ChainPasswordEncoder extends BasePasswordEncoder implements SelfSaltingEncoderInterface
+{
+    private $bestEncoder;
+    private $extraEncoders;
+
+    public function __construct(PasswordEncoderInterface $bestEncoder, PasswordEncoderInterface ...$extraEncoders)
+    {
+        $this->bestEncoder = $bestEncoder;
+        $this->extraEncoder = $extraEncoder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encodePassword($raw, $salt)
+    {
+        return $this->bestEncoder->encodePassword($raw, $salt);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        if ($this->bestEncoder->isPasswordValid($encoded, $raw, $salt)) {
+            return true;
+        }
+
+        foreach ($this->extraEncoders as $encoder) {
+            if ($encoder->isPasswordValid($encoded, $raw, $salt)) {
```

Comment:
wouldn't this cause issues in case the password hash is not sufficient to validate things because it does not embed the info about which algo was used ? Modern APIs (`password_hash` and `sodium`) are embedding that info, but older ones (see the MessageDigestPasswordEncoder for instance) are not.

I think allowing any encoder to try to validate a password hashed by another encoder could create a security risk, depending on what the encoder is doing. Especially if you include `plaintext` in the extra encoders (as that means that leaking a argon2id hash is now as bad as storing password in cleartext as they will be accepted by this validation).
For separate encoders, we should keep the logic relying on EncoderAwareInterface to choose a legacy encoder (but then, we should build a way to upgrade that to the best encoder).

Reply:
and yes, I know that always embedded the algo in the password hash would be much better, but that was not the common way years ago when we started Symfony 2.0

Reply:
isPasswordValid do validates the password by using whatever algo each encoder is using.
This means using it is sufficient to me, as we already trust it to do authentication.
I added a note about the plain text encoder.

Reply:
my issue here is that you try *all* encoders here, while old-gen implementations cannot recognizes where the hash is one using their algo, and so could accept password for a hash that is not theirs if they can generate such hash (which is why the PainTextEncoder is a huge issue, as it can generate anything as hash as it returns the input).

Reply:
All encoder implementations we have are safe in this regard. Past these, it's not our responsibility.
I added a warning about using `PlaintextPasswordEncoder` in the docblock of `ChainPasswordEncoder`.

Reply:
is MessageDigestPasswordEncoder actually safe in this regard ?

Reply:
yep, because it hashes the password and compares the hashes. Providing a non supported hash cannot collide with the password's hash.